XDMF

KyleVaughn opened the pull request:
Adds XDMF mesh type and file type. Allows for reading and writing of 2D XMDF mesh files.

KyleVaughn commented:
Note that XML element members were made public because the PRIVATE attribute was applied selectively, leaving things like `content` and `name` public despite having `setName` or `getContent` and because the setters and getters are largely unnecessary. For instance, `getParent` is just a call to `parent => thisXMLE%parent` and `getContent` is just `content=thisXMLE%content`.

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+  CHARACTER(LEN=*),PARAMETER :: myName='export_leaf_XDMFFileType'
+  TYPE(XDMFMeshType),INTENT(IN)  :: mesh
+  TYPE(XMLElementType),TARGET,INTENT(INOUT) :: xmle
+  TYPE(StringType),INTENT(INOUT) :: strpath
```
Can this be INTENT(IN)?

KyleVaughn replied:
Strpath can be. Changed

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+
+  ! Determine number of xmle children
+  ! Geometry, Topology, Material, Cell sets
+  ! Assumes mesh has geometry and topology
```
Add DBC statement?

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+PUBLIC :: XDMFMeshType
+PUBLIC :: XDMFTopologyList
+PUBLIC :: ASSIGNMENT(=)
+
```
I see a lot of indexing with 1, 2, 3, 4 integers which can be confusing when newly looking at the code. Can you assign named integers here so that more context is given for the indexing? I.e. GEOMETRY_IDX=1, TOPOLOGY_IDX=2, etc

KyleVaughn replied:
Added where possible. A lot of the 1 indexing is due to the assumption that there is only one child. There are checks for this, but now that I have it as error handling it should be more clear that the assumption exists.

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+  ALLOCATE(xmle%children(nchildren))
+
+  ! GEOMETRY
+  current_xml => xmle%children(1)
```
GEOMETRY_IDX = 1

KyleVaughn replied:
Added

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+  str_value = "XYZ"
+  CALL current_xml%setAttribute(str_name, str_value)
+
+  ALLOCATE(current_xml%children(1))
```
GEOMETRY_IDX

KyleVaughn replied:
This one is actually the allocation of an xml element with one child, hence the 1. It will always only have 1 child, so GEOMETRY_IDX doesn't work here even if they are equivalent

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+  str_name="Dimensions"
+  str1 = nverts
+  str2 = "3"
+  str_value = str1//" "//str2
```
Can this be consolidated to one line without str1 and str2?

KyleVaughn replied:
Not without changing to a formatted write statement into a character array. The str1 = nverts is an assignment from integer to string. Strings have the // operator defined, so this was easiest for me.

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+  CALL h5%fwrite(CHAR(mesh%name)//'->vertices',mesh%vertices)
+
+  ! TOPOLOGY
+  current_xml => xmle%children(2)
```
TOPOLOGY_IDX

KyleVaughn replied:
Added

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+
+    str_name="Dimensions"
+    str1 = ncells
+    str2 = nverts
```
nverts is set to str1 at line 879. can set str2 to ncells instead of reassigning both.

KyleVaughn replied:
True, but 50 lines is a ways away and assignment is extremely fast, so I think this makes sense.

thfolk commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+        CALL xmle_children(i)%getChildren(ele_children)
+        REQUIRE(SIZE(ele_children) == 1)
+        strIn='Format'
+        CALL ele_children(1)%getAttributeValue(strIn,strOut)
```
Instead of indexing children(1) throughout, add a named INTEGER. Is (1) the parent or a top level? Could do PARENT_IDX.

KyleVaughn replied:
The (1) refers to the only child in a lot of places. I feel (1) is better than (ONLY_CHILD) or something like that.

thfolk reviewed (commented):
Hey Kyle, overall looks great! Awesome job. I followed it pretty well for the most part. Just a few things I think could improve the readability.
1) Used named integers instead of indexing with 1, 2, 3, etc. Examples include GEOMETRY_IDX, TOPOLOGY_IDX, etc. for your indexes that are fixed. 
2) Most of the variables were descriptive. Some were not. I suggested a few changes to some variables that could show more context.
3) There are a lot of DBC statements filtered throughout. It's good that you are checking things as you go. From what I've learned about DBC statements they normally go at the beginning of a module. Some of your statements could be turned into IF blocks with an exception handler. Since there are so many DBC statements throughout, you could create a module to check all of these things in one place. 

Overall, great work!

aarograh commented:
@thfolk @KyleVaughn I'm seeing a couple comments about DBC vs. exception handlers.  First off, thanks for adding one or the other... something is always better than nothing!  I figured I'd take a minute to write a note about picking between the two though in case it's relevant to the discussion going on here:

* Exception Handling: Use this for bad conditions that could potentially be caused by user input.  For example, two input options are related to each other: input A can only have certain values depending on what value input B has.  That situations warrants a descriptive error message to provide feedback to the user.  We often can't reasonably test all input combinations during input processing, so we need to provide feedback downstream too.
* DBC: I think of this more as feedback to the developer.  In other words, if a DBC statement trips, then that probably means the developer called the routine in a way that it should never be called.  It should be rare for a user to be able to cause a DBC statement to appear, since "Design by Contract" really refers to how we as developers tie various functions and classes together.  If the developer obeyed the "contract", the user should never see this.  This is also partly why we disable DBC checks in release builds, since they shouldn't be encountered by users anyway.

I haven't looked in detail at this code so I don't know which of these is appropriate (maybe both are appropriate in different places), but hopefully that helps in deciding whether or not things should be changed from one to the other.

KyleVaughn commented:
Thanks @aarograh @thfolk !
It looks like I should change a decent amount of the DBC to exception handling, since it's primarily applied to assumptions about valid XMDF, valid HDF5, and what structure the mesh file will have.

thfolk commented:
Thanks @aarograh. I'm saving your comment to keep in the memory bank!

KyleVaughn commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+  TYPE(HDF5FileType), INTENT(INOUT) :: h5
+  TYPE(XMLElementType), POINTER :: xmle_children(:)
+  TYPE(StringType) :: strIn, strOut
+  INTEGER(SIK) :: i, grid_ctr, mesh_ctr
```
Changed

KyleVaughn commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+        ENDIF
+        ! Format
+        CALL xmle_children(i)%getChildren(ele_children)
+        REQUIRE(SIZE(ele_children) == 1)
```
Changed to error handler

KyleVaughn commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+        strIn='Dimensions'
+        CALL ele_children(1)%getAttributeValue(strIn,strOut)
+        strArray=strOut%split()
+        REQUIRE(strArray(2) == '3')
```
Changed to error handler. Also, the error checking happens as each component is read in. The code below double checks that the h5 and xdmf data match and are the correct shape.

KyleVaughn commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+        group = group%replace(NEW_LINE("A"),"")
+        REQUIRE(h5%pathExists(CHAR(group)))
+        group = group%replace("/", "->")
+        ! Data shape
```
Fixed string manipulation to be more readable.

KyleVaughn commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+  IF (SIZE(children) == 2) THEN
+    REQUIRE(children(1)%name%upper() == 'INFORMATION')
+    REQUIRE(children(2)%name%upper() == 'GRID')
+    i = 2
```
Changed

KyleVaughn commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1236 @@
+    ENDDO
+    DEALLOCATE(thismesh%cell_sets)
+  ENDIF
+ENDSUBROUTINE clear_XDMFMeshType
```
No, but I checked that every member is deallocated from bottom up, so the only memory that should stick around is the memory needed to hold the uninitialized mesh object. Testing in Valgrind seems like overkill when the clear routine is only 20 lines.

KyleVaughn commented:
I believe the tests are failing now due to CASL Trilinos being moved.
```
git clone https://github.com/CASL/Trilinos.git
Cloning into 'Trilinos'...
remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/CASL/Trilinos.git/'
The command "git clone https://github.com/CASL/Trilinos.git" exited with 128.
```

yl5 commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1324 @@
+!Import and export are the only procedures used.
+!open, close, and delete are members of the base type.
+  CONTAINS
+    !> @copybrief FileType_XDMF::fopen_XDMFFileType
+    !> @copydoc FileType_XDMF::fopen_XDMFFileType
+    PROCEDURE,PASS :: fopen => fopen_XDMFFileType
+    !> @copybrief FileType_XDMF::fclose_XDMFFileType
+    !> @copydoc FileType_XDMF::fclose_XDMFFileType
+    PROCEDURE,PASS :: fclose => fclose_XDMFFileType
+    !> @copybrief FileType_XDMF::fdelete_XDMFFileType
+    !> @copydoc FileType_XDMF::fdelete_XDMFFileType
+    PROCEDURE,PASS :: fdelete => fdelete_XDMFFileType
```
Is any information used from the BaseFileType? If not, I don't see the XDMF should be derived from it. Right now, you have to implement these deferred procedures (and they don't do anything). Maybe XDMF is a data structure or format that is not suitable to be defined as a file type?

yl5 commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1324 @@
+  CLASS(XDMFFileType),INTENT(INOUT) :: thisXDMFFile
+  TYPE(StringType),INTENT(INOUT) :: strpath
+  TYPE(XDMFMeshType),INTENT(OUT),TARGET  :: mesh
+  TYPE(XMLFileType) :: xml
+  TYPE(HDF5FileType) :: h5
```
Looks like you are reading in multiple files to construct the XDMF data structure so that suggests again it shouldn't be derived from the BaseFileType.

yl5 commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1324 @@
+!-------------------------------------------------------------------------------
+!> @brief Gets the number of grid levels to a leaf node
+!> @param thismesh the XDMF mesh object
+!>
```
Add return variable to comment, e.g.,
!> @returns n blabla...

yl5 commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1324 @@
+  INTEGER(SNK) :: n
+  ! It is assumed that all siblings have children if any sibling has children
+  ! Hence it is sufficient to assess thismesh%children(1)'s depth recursively
+  ! Ex:     Possible                Not Possible
```
Nice mini figure:-)

yl5 commented on src/FileType_XDMF.f90:
```diff
@@ -0,0 +1,1324 @@
+  IF(.NOT.XDMFTopologyList%has('Topology')) CALL init_XDMFTopologyList()
+
+  !H5
+  ! NOTE: it is assumed that the h5 and xml files have the same name.
```
Maybe better to add an exception if prefer to enforce user doing this.

yl5 commented on src/FileType_XML.f90:
```diff
@@ -55,19 +55,19 @@ MODULE FileType_XML
 !> Derived type for an XML element
 TYPE :: XMLElementType
   !> The number of attributes defined on the element
-  INTEGER(SIK),PRIVATE :: nAttr=0
```
If they were set to private for some purpose (e.g., avoid accidentally overriding it), it should be better to write setter/getter to access them. Same thing applies to the below changes.

yl5 commented on unit_tests/testFileType_XDMF/testFileType_XDMF.f90:
```diff
@@ -0,0 +1,618 @@
+IMPLICIT NONE
+
+REAL(SDK) :: two_pins_pin1_vertices(3,109) = RESHAPE( (/ &
+    0.0000000000000000,        0.0000000000000000,        0.0000000000000000,&
```
need _SDK otherwise they will be single precision

yl5 commented on unit_tests/testFileType_XDMF/testFileType_XDMF.f90:
```diff
@@ -0,0 +1,618 @@
+
+
+CREATE_TEST('XDMF TYPE')
+REGISTER_SUBTEST('two_pins',test_two_pins)
+REGISTER_SUBTEST('three_level_grid',test_three_level_grid)
+REGISTER_SUBTEST('three_level_grid_IH', test_three_level_grid_implicit_hierarchy)
```
SUBTEST should be to test each procedure (constructor/destructor, type-bound procedure, etc.) rather than each type of problems. COMPONENT_TEST may be used to test each case/problem.

yl5 commented:
Overall looks good, but I'd suggest to reconsider deriving the type from the BaseFileType and rewriting the unit tests.

KyleVaughn commented:
@yl5 Addressed all of your comments except for making XML element's children attribute private again. I need XML elements to persist in memory after leaving a subroutine, so making an allocatable XML element array, setting a pointer to it, then using the XML setChildren method isn't sufficient, since it just does `thisXMLE%children => children` and the underlying XML elements will be deallocated once leaving the subroutine, so the pointer persists longer than the target. I could potentially write a function just to allocate the children, then use getChildren to work with children one by one, but this goes back to the point of unnecessary functions. I don't think that hiding these attributes makes sense, since the XML class does not have any derived classes, so there is no benefit to writing setters, getters, etc. to provide a general interface to the object. I'm fine with leaving the other attributes private, since they have the methods they need to be useful for the most part, but not being able to access children is a real hindrance to working with the type.

HendersonSC commented:
> @yl5 Addressed all of your comments except for making XML element's children attribute private again. I need XML elements to persist in memory after leaving a subroutine, so making an allocatable XML element array, setting a pointer to it, then using the XML setChildren method isn't sufficient, since it just does `thisXMLE%children => children` and the underlying XML elements will be deallocated once leaving the subroutine, so the pointer persists longer than the target. I could potentially write a function just to allocate the children, then use getChildren to work with children one by one, but this goes back to the point of unnecessary functions. I don't think that hiding these attributes makes sense, since the XML class does not have any derived classes, so there is no benefit to writing setters, getters, etc. to provide a general interface to the object. I'm fine with leaving the other attributes private, since they have the methods they need to be useful for the most part, but not being able to access children is a real hindrance to working with the type.

I would claim that there are multiple reasons for a setter/getter and there are many reasons for one to exists. Further its existence and/or relevance is wholly unrelated to its internal logic (or lack thereof).

To the benefit of a developer some `setFoo` or `getBar` tied to a private data component, provides an absolute guarantee that you know where the data is changing if it changes. Since you know where it changes, during debugging sessions you can know when it changes, which inevitably leads to why it changes. Which makes things way easier than hunting for where a variable changes.

Take note, that I haven't reviewed the underlying code, just agreeing with @yl5. It is plausible that it needs to be public, but the reason it needs to be public can not be that the appropriate setter/getter is too simple. If it is just the technical problem of a pointer outliving target or I have somehow missed something, feel free to email me and/or ping me on slack.

KyleVaughn commented:
@yl5 Talked with Brendan, and I'm going to keep it private. Will update later today.

yl5 commented:
@KyleVaughn, just a couple things as I saw your updates:

1) If you agree that XDMF is not a file type, maybe change the module and source file names as well. Maybe XMDFDataType or XMDFDataLibrary?

2) I might not be clear about the unit tests. The purpose is to test each subroutine independently to make sure we get expected outputs from the input parameters. We should write a SUBTEST for each type-bound or public procedure as the minimum. Within each SUBTEST, you can test different problem conditions (by varying the input parameters) to thoroughly cover the codes in that procedure, and this is where you may use the COMPONENT_TEST. The current implementation seems to cover the ballpark of the testing but you will need to breakdown/restructure it to test each procedure independently. For an integration test of multiple routines and the overall feature, that is normally what a regression test does.

KyleVaughn commented:
@yl5 Renamed the module to XDMFMesh and removed the XDMFFileType, so import and export are simply importXDMFMesh and exportXDMFMesh. 

The unit tests were rewritten to maintain independence as best as possible. Functions are used in other tests once that test has passed, like the clear routine and assignment operator. The export tests use the import subroutine to verify, which I'm not happy about, but setting up all the meshes that should be tested by hand would be tedious and a lot of code. Importing the exported meshes also allows me to access the written data easier. 

The import and export now have component tests for the various cases that test the major branches in logic, as denoted by the table in each function. 

Lastly, the XML children attribute was made private again.

yl5 reviewed (approved):
Comments were addressed properly. Looks good to me now.
